src/backend/libpq/auth.c

```diff
@@ -3256,57 +3261,89 @@ CheckAuthTimeConstraints(char *rolname)
+	HeapTuple		tuple;
+	authPoint 		now;
+	bool			authorized = true;
+
+	timestamptz_to_point(timestamp, &now);
+
 	/*
-	 * GPDB_90_MERGE_FIXME: get_role_intervals() no longer exists. We need to
-	 * walk pg_auth_time_constraint for this information now, instead of relying
-	 * on a flat file.
+	 * Disable immediate interrupts while doing database access.  (Note
+	 * we don't bother to turn this back on if we hit one of the failure
+	 * conditions, since we can expect we'll just exit right away anyway.)
 	 */
-	List  	  		*role_intervals = NIL; // get_role_intervals(rolname);
+	ImmediateInterruptOK = false;
 
-	if (role_intervals == NIL)
-		return true;
+	/* Look up this user in pg_authid. */
+	roleId = GetSysCacheOid1(AUTHNAME, CStringGetDatum(rolname));
+	if (!OidIsValid(roleId))
+		return false;					/* no such user */
 
-	List   	   		*row;
-	ListCell   		*line, *cell;
-	char   	   		*temp;
-	authInterval	given;
-	authPoint 		now;
+	/* Walk pg_auth_time_constraint for entries belonging to this user. */
+	reltimeconstr = heap_open(AuthTimeConstraintRelationId, AccessShareLock);
 
-	timestamptz_to_point(timestamp, &now);
-	
-	foreach (line, role_intervals) 
+	ScanKeyInit(&entry[0],
+				Anum_pg_auth_time_constraint_authid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(roleId));
+
+	/*
+	 * Since this is called during authentication, we have to make sure we don't
+	 * use the index unless it's already been built. See GetDatabaseTuple for
+	 * another example of this sort of logic.
+	 */
+	scan = systable_beginscan(reltimeconstr, AuthTimeConstraintAuthIdIndexId,
+							  criticalSharedRelcachesBuilt, SnapshotNow, 1,
+							  entry);
+
+	/*
+	 * Check each denied interval to see if the current timestamp is part of it.
+	 */
+	while (HeapTupleIsValid(tuple = systable_getnext(scan)))
 	{
-		row = lfirst(line);
-		/* 
-		 * Each row of role_intervals is a List of 5 records
-		 * <rolname> <startday> <starttime> <endday> <endtime> 
-	 	 */
-		cell = list_head(row);
-		cell = lnext(cell);			/* skip first entry in record, which is rolname */
-		temp = lfirst(cell);
-		given.start.day = pg_atoi(temp, sizeof(int16), 0);
-
-		cell = lnext(cell);
-		temp = lfirst(cell);
-		given.start.time = DatumGetTimeADT(DirectFunctionCall1(time_in,
-										   CStringGetDatum(temp)));
-		
-		cell = lnext(cell);
-		temp = lfirst(cell);
-		given.end.day = pg_atoi(temp, sizeof(int16), 0);
-	
-		cell = lnext(cell);
-		temp = lfirst(cell);
-		given.end.time = DatumGetTimeADT(DirectFunctionCall1(time_in,
-										 CStringGetDatum(temp)));
+		Form_pg_auth_time_constraint constraint_tuple;
+		Datum 			datum;
+		bool			isnull;
+		authInterval	given;
+
+		constraint_tuple = (Form_pg_auth_time_constraint) GETSTRUCT(tuple);
+		Assert(constraint_tuple->authid == roleId);
+
+		/* Retrieve the start of the interval. */
+		datum = heap_getattr(tuple, Anum_pg_auth_time_constraint_start_time,
+							 RelationGetDescr(reltimeconstr), &isnull);
+		Assert(!isnull);
```

Comment:
Is this an appropriate check (there's a non-null constraint on that column), or should I just not worry about it and remove the out-param entirely?

Reply:
An assertion seems reasonable to me. I would not expect it to ever fail, but it's a useful nugget of documentation, pointing out that it can never be null.